_posts/2022-06-08-#22778.md

```diff
@@ -0,0 +1,109 @@
+---
+layout: pr
+date: 2022-06-08
+title: "net processing: Reduce resource usage for inbound block-relay-only connections"
+pr: 22778
+authors: [jnewbery]
+components: ["p2p"]
+host: larryruane
+status: upcoming
+commit: 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4
+---
+
+## Notes
+
+- (These notes describe behavior with or without this PR.)
+
+- We are a `bitcoind` node. Each of our peers is either inbound (our
+  peer initiated the connection) or outbound (we initiated the connection).
+  Once the connection is established, the relationship is symmetric (the
+  protocol works the same in both directions).
+
+- When we initiate an outbound connection, we may decide that we won't
+  relay (forward) transactions to this peer.
+  This peer is designated a `block-relay-only` peer. The peer
+  doesn't know that we're treating it as `block-relay-only`.
+
+- When connecting with a peer, we allocate a `Peer` object.
+  If we'll need to notify and send transactions to
+  this peer (for example, if it's not a `block-relay-only` peer),
+  we create an additional `TxRelay` object to keep track
+  of state needed for relaying transactions to this peer.
+
+- The `TxRelay` object contains (in simplified form):
+   - the list of transactions (txids) that we know about
+     but haven't yet notified this peer about
+   - the "list" (actually a bloom filter) of transactions (txids) that
+     the peer knows about (because either we sent the transaction
+     to the peer, or it has sent it to us)
+   - if we've enabled BIP37, the bloom filter that the peer sends us
+     so that we can filter out transactions that it doesn't care about, 
+  
+- If we advertise support for BIP37 (bloom filter) to our peer
+  during version handshake, the peer can send us a bloom filter
+  after which we won't send transactions that aren't present in
+  the filter (to reduce bandwidth).
+
+- If a node connects to us (it's one of our inbound connections), it
+  can indicate that it doesn't want to receive transactions
+  from us. It does this using the `relay` field in the
```

Comment:
I don't recommend linking to the wiki; it's often outdated and inaccurate. Linking to the code is better! https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/net_processing.cpp#L2763-L2764